subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainDependencyToComponentIdResolver.java

```diff
@@ -46,18 +48,25 @@ public void add(ModuleComponentRepository repository) {
 
     public void resolve(DependencyMetadata dependency, ModuleIdentifier targetModuleId, BuildableComponentIdResolveResult result) {
         ModuleVersionSelector requested = dependency.getRequested();
-        VersionSelector versionSelector = versionSelectorScheme.parseSelector(requested.getVersion());
-        if (versionSelector.isDynamic()) {
-            dynamicRevisionResolver.resolve(dependency, versionSelector, result);
+        VersionConstraintInternal constraint = (VersionConstraintInternal) requested.getVersionConstraint();
+        VersionSelector preferredSelector = constraint.getPreferredSelector(versionSelectorScheme);
+        if (preferredSelector.isDynamic()) {
+            dynamicRevisionResolver.resolve(dependency, preferredSelector, result);
         } else {
-            DefaultModuleComponentIdentifier id = new DefaultModuleComponentIdentifier(requested.getGroup(), requested.getName(), requested.getVersion());
+            String version;
+            if (preferredSelector instanceof ExactVersionSelector) {
+                version = ((ExactVersionSelector) preferredSelector).getSelector();
+            } else {
+                version = requested.getVersion();
+            }
+            DefaultModuleComponentIdentifier id = new DefaultModuleComponentIdentifier(requested.getGroup(), requested.getName(), version);
             ModuleVersionIdentifier mvId = moduleIdentifierFactory.moduleWithVersion(targetModuleId, requested.getVersion());
             result.resolved(id, mvId);
         }
         if (result.hasResult()) {
-            result.setVersionSelector(versionSelector);
+            VersionSelector rejectSelector = constraint.getRejectionSelector(versionSelectorScheme);
+            result.setSelectors(preferredSelector, rejectSelector);
```

Comment:
It might be nicer to retain the original constraint here, so change this to `request.setVersionConstraint(constraint)`.